pycqed/measurement/calibration/single_qubit_gates.py

```diff
@@ -328,41 +339,152 @@ def sweep_block(self, qb, sweep_points, flux_op_code=None,
+                            'with the auto mode of ro_pulse_delay.')
+
+        assert not (fp_compensation and fp_during_ro)
+
         pulse_modifs = {'attr=name,op_code=X180': f'FPS_Pi',
                         'attr=element_name,op_code=X180': 'FPS_Pi_el'}
         b = self.block_from_ops(f'ge_flux {qb}',
-                                [f'X180 {qb}', flux_op_code],
+                                [f'X180 {qb}'] + [flux_op_code] * \
+                                (2 if fp_compensation else 1) \
+                                + ([f'FP {qb}'] if fp_during_ro else []),
                                 pulse_modifs=pulse_modifs)
+
         fp = b.pulses[1]
         fp['ref_point'] = 'middle'
-        offs = fp.get('buffer_length_start', 0)
+        bl_start = fp.get('buffer_length_start', 0)
+        bl_end = fp.get('buffer_length_end', 0)
+
+        def fp_delay(x, o=bl_start):
+            return -(x+o)
+
         fp['pulse_delay'] = ParametricValue(
-            'delay', func=lambda x, o=offs: -(x + o))
+            'delay', func=fp_delay)
+
+        fp_length_function = lambda x: fp['pulse_length']
+
+        if (fp_truncation or hasattr(fp_truncation, '__iter__')):
+            if not hasattr(fp_truncation, '__iter__'):
+                fp_truncation = [-np.inf, np.inf]
+            original_fp_length = fp['pulse_length']
+            max_fp_sweep_length = np.max(
+                sweep_points.get_sweep_params_property(
+                    'values', dimension=0, param_names='delay'))
+            sweep_diff = max(max_fp_sweep_length - original_fp_length, 0)
+            fp_length_function = lambda x, opl=original_fp_length, \
+                o=bl_start + fp_truncation_buffer, trunc=fp_truncation: \
+                max(min((x + o), opl), 0) if (x>np.min(trunc) and x<np.max(trunc)) else opl
+            # TODO: check what happens if buffer_length_start and buffer_length_end are zero.
+
+            fp['pulse_length'] = ParametricValue(
+                'delay', func=fp_length_function)
+            if fp_compensation:
+                cp = b.pulses[2]
+                cp['amplitude'] = -np.sign(fp['amplitude']) * np.abs(
+                    fp_compensation_amp)
+                cp['pulse_delay'] = sweep_diff + bl_start
+                tau = 200e-9 * 100
+
+                def t_trunc(x, fnc=fp_length_function, tau=tau,
+                            fp_amp=fp['amplitude'], cp_amp=cp['amplitude']):
+                    fp_length = fnc(x)
+
+                    def v_c(tau, fp_length, fp_amp, v_c_start=0):
+                        return fp_amp - (fp_amp - v_c_start) * np.exp(
+                            -fp_length / tau)
+
+                    v_c_fp = v_c(tau, fp_length, fp_amp, v_c_start=0)
+                    return -np.log(cp_amp / (cp_amp - v_c_fp)) * tau
+
+                cp['pulse_length'] = ParametricValue('delay', func=t_trunc)
+                # TODO: implement that the ro_delay is adjusted accordingly!
```

Comment:
will this lead to problems in the future? Or maybe already in some particular cases?

Reply:
Fixed the fp_compensation feature to reference the RO pulse to the end of the custom compensation pulse. This way, the same ro_delay can be used.